src/firebaseConfig.ts

```diff
@@ -61,18 +62,58 @@ export type HostingConfig = {
   };
 };
 
-export type StorageConfig =
-  | {
-      rules: string;
-    }
-  | {
-      bucket: string;
-      rules: string;
-    }[];
+type HostingOne = HostingBase &
+  DeployAsset & {
+    site?: string;
+    target?: string;
+  };
+
+type HostingMany = (HostingBase &
+  (
+    | {
+        site: string;
+        target?: string;
+      }
+    | {
+        site?: string;
+        target: string;
+      }
+  ) &
+  DeployAsset)[];
+
+type StorageOne = {
+  rules: string;
+  target?: string;
+} & DeployAsset;
+
+type StorageMany = ({
+  rules: string;
+  bucket: string;
+  target?: string;
+} & DeployAsset)[];
+
+// Full Configs
+export type DatabaseConfig = DatabaseOne | DatabaseMany;
+
+export type FirestoreConfig = {
+  rules?: string;
+  indexes?: string;
+} & DeployAsset;
+
+export type FunctionsConfig = {
+  // TODO: Add types for "backend"
+  source?: string;
+  ignore?: string[];
+  runtime?: string;
```

Comment:
Should this be an list of valid strings? Is there a central list this can be sourced from?

Reply:
I tried using:

```diff
+import * as runtimes from "./deploy/functions/runtimes";

export type FunctionsConfig = {
  // TODO: Add types for "backend"
  source?: string;
  ignore?: string[];
-  runtime?: string;
+  runtime?: runtimes.Runtime;
} & DeployAsset;
```

But the `npm run generate:json-schema` function produces errors because of the import:

<details>
<summary>errors</summary>s

```
/home/jthegedus/projects/fork-firebase-tools/node_modules/agent-base/dist/src/index.d.ts (2,8): Module '"net"' has no default export.
/home/jthegedus/projects/fork-firebase-tools/node_modules/agent-base/dist/src/index.d.ts (3,8): Module '"http"' has no default export.
/home/jthegedus/projects/fork-firebase-tools/node_modules/agent-base/dist/src/index.d.ts (4,8): Module '"https"' has no default export.
/home/jthegedus/projects/fork-firebase-tools/src/apiv2.ts (7,8): Module '"util"' has no default export.
/home/jthegedus/projects/fork-firebase-tools/src/defaultCredentials.ts (79,5): Type 'string | undefined' is not assignable to type 'undefined'.
  Type 'string' is not assignable to type 'undefined'.
/home/jthegedus/projects/fork-firebase-tools/src/defaultCredentials.ts (83,7): Type 'string' is not assignable to type 'undefined'.
/home/jthegedus/projects/fork-firebase-tools/src/gcp/proto.ts (32,20): Argument of type 'keyof T' is not assignable to parameter of type 'never'.
  Type 'string | number | symbol' is not assignable to type 'never'.
    Type 'string' is not assignable to type 'never'.
/home/jthegedus/projects/fork-firebase-tools/src/prompt.ts (37,20): Argument of type 'Question' is not assignable to parameter of type 'never'.
  Type 'ListQuestion<Answers>' is not assignable to type 'never'.
/home/jthegedus/projects/fork-firebase-tools/src/prompt.ts (53,15): No overload matches this call.
  Overload 1 of 2, '(o: {}): string[]', gave the following error.
    Argument of type 'unknown' is not assignable to parameter of type '{}'.
  Overload 2 of 2, '(o: object): string[]', gave the following error.
    Argument of type 'unknown' is not assignable to parameter of type 'object'.
/home/jthegedus/projects/fork-firebase-tools/src/prompt.ts (54,18): Object is of type 'unknown'.
/home/jthegedus/projects/fork-firebase-tools/node_modules/typescript-json-schema/dist/typescript-json-schema.js:1290
                throw new Error("No output definition. Probably caused by errors prior to this?");
```

</details>

Reply:
Updated with the string literals of currently available runtimes